components/net/http_cache.rs

```diff
@@ -857,7 +871,32 @@ impl HttpCache {
                 last_validated: time::now(),
             }),
         };
+        debug!("storing new cached response for {:?}", request.url());
         let entry = self.entries.entry(entry_key).or_insert(vec![]);
+
+        if response
+            .status
+            .as_ref()
+            .map_or(false, |s| s.0 == StatusCode::OK)
+        {
+            // Ensure that any existing complete response is overwritten by the new
+            // complete response.
+            let existing_complete_response = entry.iter().position(|response| {
```

Comment:
I'm less sure about this, since in theory different complete responses could be cached and used to satisfy different requests, for example in the case of request with different `Vary` headers. Have you tried running this test with those changes: https://github.com/servo/servo/blob/2313f7f37002981cedae4b58bf90e71d6044fb83/tests/wpt/web-platform-tests/fetch/http-cache/vary.html#L77 ?

So its not clear to me what the underlying problem is with the noted behavior of "old cached content could be viewed as a result of revalidating a newer cached response". If "old" is still valid, there are cases where the older content should be constructed versus a newer cached response. 

Also what do you mean with "as a result of revalidating a newer cached response"? If an otherwise expired cached resource is refreshed when it should not, it might be because the refresh logic is not yet fully spec compliant in terms of selecting a resource based on strong/weak validators as described at https://tools.ietf.org/html/rfc7234#section-4.3.4